### PR TITLE
2FA for 'Approve' action in RFM

### PR DESCRIPTION
**Pull request by Ehtesham599**
## Feature description
As an Approver for the purchasing process, I want to approve with a 2 factor authentication method, so I can secure the approval and signature process.

## Solution description
- Implemented server side 2FA in common utils which can be used for any doctype.
- The verification code is a random 6-digit number.
- Verification code is stored is stored in session cache.
- Verification code has a timeout of 5 minutes (300s).
- Verification code is sent via email to logged in user when an 'Approve' action takes place.
- Implemented client side dialog box to enter verification code with an option to resend.
- Added client side validation and error handling.

## Output screenshots (optional)
Upon clicking 'approve' button:
Screen Shot 2022-01-09 at 10 41 35 AM
Screen Shot 2022-01-09 at 10 42 11 AM

Email received at user end:
Screen Shot 2022-01-09 at 10 44 03 AM

Signature Fetched from Employee Doc:
Screen Shot 2022-01-09 at 3 48 14 PM

## Areas affected and ensured
Nothing affected.

## Is there any existing behavior change of other features due to this code change?
Yes. Approval will now require a two factor authentication in the RFM workflow.

## Was this feature tested on all the browsers?
  - [x] Chrome
  - [x] Safari

### CODE REVIEW

**Comment by Ehtesham599** [comment]
@SalShai You can refer to this implementation in RFM and use the same for any DocTypes that require signature for Approval.

File: one_fm/purchase/doctype/request_for_purchase/request_for_purchase.js [hunk 1]
```diff
@@ -134,6 +185,7 @@ frappe.ui.form.on('Request for Purchase', {
 		});
 	},
 	make_purchase_order: function(frm) {
+		console.log(frm.doc)
```
Comment (pjamsheer) [inline one_fm/purchase/doctype/request_for_purchase/request_for_purchase.js:188]: Remove consol log

**Comment by pjamsheer** [comment]
@Ehtesham599 Please resolve the conflict
@Ehtesham599 and @SalShai I have a suggestion, Can we Optimise the verification client side code in a controller and make use of it in all the doctype ??

**Comment by Ehtesham599** [comment]
> @Ehtesham599 Please resolve the conflict @Ehtesham599 and @SalShai I have a suggestion, Can we Optimise the verification client side code in a controller and make use of it in all the doctype ??

@pjamsheer Yeah I think we could do that. I just implemented for one doctype for reference only.

**Comment by Ehtesham599** [comment]
@SalShai In PO and PR please update frappe.confirm() message. They have "Request for Material" mentioned in them.

**Comment by pjamsheer** [comment]
@SalShai resolve the conflict by rebase with staging

**Comment by pjamsheer** [comment]
> > @Ehtesham599 Please resolve the conflict @Ehtesham599 and @SalShai I have a suggestion, Can we Optimise the verification client side code in a controller and make use of it in all the doctype ??
> 
> @pjamsheer Yeah I think we could do that. I just implemented for one doctype for reference only.

@Ehtesham599 I hope for now we can go with this change and later we can do the optimisation as a chore !!

**Review by pjamsheer — APPROVED** [review]

**Pull request merged**